#Librerías necesarias para la realización del TP

In [ ]:
import pandas as pd


#Dataset necesario para la realización del TP

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
try:
  # el archivo ya está cargado en el sistema
  csv = pd.read_csv('vehicles_dataset.csv')
except:
  # cargar usando el acceso directo 'Datasets'
  # puesto en la carpeta MyDrive del drive del usuario
  csv = pd.read_csv('drive/MyDrive/Dataset_BDD/vehicles_dataset.csv')

Analisis de Preprocesamiento

In [ ]:
df = csv.copy()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1002 entries, 0 to 1001
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            1002 non-null   object 
 1   description     946 non-null    object 
 2   make            1002 non-null   object 
 3   model           1002 non-null   object 
 4   type            1002 non-null   object 
 5   year            1002 non-null   int64  
 6   price           979 non-null    float64
 7   engine          1000 non-null   object 
 8   cylinders       897 non-null    float64
 9   fuel            995 non-null    object 
 10  mileage         968 non-null    float64
 11  transmission    1000 non-null   object 
 12  trim            1001 non-null   object 
 13  body            999 non-null    object 
 14  doors           995 non-null    float64
 15  exterior_color  997 non-null    object 
 16  interior_color  964 non-null    object 
 17  drivetrain      1002 non-null   o

In [ ]:
# Ver las primeras 5 filas del DataFrame
df.head(5)


name  \
0     2024 Jeep Wagoneer Series II   
1  2024 Jeep Grand Cherokee Laredo   
2         2024 GMC Yukon XL Denali   
3       2023 Dodge Durango Pursuit   
4            2024 RAM 3500 Laramie   

                                         description   make           model  \
0  \n      \n        Heated Leather Seats, Nav Sy...   Jeep        Wagoneer   
1  Al West is committed to offering every custome...   Jeep  Grand Cherokee   
2                                                NaN    GMC        Yukon XL   
3  White Knuckle Clearcoat 2023 Dodge Durango Pur...  Dodge         Durango   
4  \n      \n        2024 Ram 3500 Laramie Billet...    RAM            3500   

  type  year    price                                             engine  \
0  New  2024  74600.0                            24V GDI DOHC Twin Turbo   
1  New  2024  50170.0                                                OHV   
2  New  2024  96410.0  6.2L V-8 gasoline direct injection, variable v...   
3  New  2023  46835.0                                       16V MPFI OHV   
4  New  2024  81663.0                           24V DDI OHV Turbo Diesel   

   cylinders      fuel  mileage       transmission       trim          body  \
0        6.0  Gasoline     10.0  8-Speed Automatic  Series II           SUV   
1        6.0  Gasoline      1.0  8-Speed Automatic     Laredo           SUV   
2        8.0  Gasoline      0.0          Automatic     Denali           SUV   
3        8.0  Gasoline     32.0  8-Speed Automatic    Pursuit           SUV   
4        6.0    Diesel     10.0  6-Speed Automatic    Laramie  Pickup Truck   

   doors           exterior_color    interior_color        drivetrain  
0    4.0                    White      Global Black  Four-wheel Drive  
1    4.0                 Metallic      Global Black  Four-wheel Drive  
2    4.0             Summit White  Teak/Light Shale  Four-wheel Drive  
3    4.0  White Knuckle Clearcoat             Black   All-wheel Drive  
4    4.0                   Silver             Black  Four-wheel Drive

In [ ]:
df.describe()

year          price   cylinders      mileage       doors
count  1002.000000     979.000000  897.000000   968.000000  995.000000
mean   2023.916168   50202.985700    4.975474    69.033058    3.943719
std       0.298109   18700.392062    1.392526   507.435745    0.274409
min    2023.000000       0.000000    0.000000     0.000000    2.000000
25%    2024.000000   36600.000000    4.000000     4.000000    4.000000
50%    2024.000000   47165.000000    4.000000     8.000000    4.000000
75%    2024.000000   58919.500000    6.000000    13.000000    4.000000
max    2025.000000  195895.000000    8.000000  9711.000000    5.000000

In [ ]:
df.nunique()

name              358
description       761
make               28
model             153
type                1
year                3
price             859
engine            100
cylinders           5
fuel                7
mileage            95
transmission       38
trim              197
body                8
doors               4
exterior_color    263
interior_color     91
drivetrain          4
dtype: int64

 La columna 'type' no aporta valor al dataset, asi que se ignora al ser un único valor (new).

In [ ]:
df.drop(columns='type', inplace=True)

#Análisis de datos nulos

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1002 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            1002 non-null   object 
 1   description     946 non-null    object 
 2   make            1002 non-null   object 
 3   model           1002 non-null   object 
 4   year            1002 non-null   int64  
 5   price           979 non-null    float64
 6   engine          1000 non-null   object 
 7   cylinders       897 non-null    float64
 8   fuel            995 non-null    object 
 9   mileage         968 non-null    float64
 10  transmission    1000 non-null   object 
 11  trim            1001 non-null   object 
 12  body            999 non-null    object 
 13  doors           995 non-null    float64
 14  exterior_color  997 non-null    object 
 15  interior_color  964 non-null    object 
 16  drivetrain      1002 non-null   object 
dtypes: float64(4), int64(1), object(1

Completamos los datos nulos de la columna 'description' con un '-'

In [ ]:
df['description'] = df['description'].fillna('-')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1002 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            1002 non-null   object 
 1   description     1002 non-null   object 
 2   make            1002 non-null   object 
 3   model           1002 non-null   object 
 4   year            1002 non-null   int64  
 5   price           979 non-null    float64
 6   engine          1000 non-null   object 
 7   cylinders       897 non-null    float64
 8   fuel            995 non-null    object 
 9   mileage         968 non-null    float64
 10  transmission    1000 non-null   object 
 11  trim            1001 non-null   object 
 12  body            999 non-null    object 
 13  doors           995 non-null    float64
 14  exterior_color  997 non-null    object 
 15  interior_color  964 non-null    object 
 16  drivetrain      1002 non-null   object 
dtypes: float64(4), int64(1), object(1

In [ ]:
df.cylinders.value_counts()

cylinders
4.0    502
6.0    277
8.0     89
3.0     27
0.0      2
Name: count, dtype: int64

In [ ]:
df[df ['cylinders'].isnull()]

name  \
14             2024 Chevrolet Blazer EV 2LT   
28             2024 Chevrolet Blazer EV 2LT   
33                          2024 Kia EV6 GT   
35         2024 Ford Mustang Mach-E Premium   
49   2024 Hyundai IONIQ 5 SE Standard Range   
..                                      ...   
884                         2024 BMW i7 M70   
893                    2025 Honda CR-V EX-L   
941                 2024 Hyundai IONIQ 5 SE   
944                       2024 Kia EV6 Wind   
978                       2024 Kia EV6 Wind   

                                           description       make  \
14   \n      \n        Sterling Gray Metallic 2024 ...  Chevrolet   
28   Radiant Red Tintcoat 2024 Chevrolet Blazer EV ...  Chevrolet   
33   Yacht Blue 2024 Kia EV6 GT AWD 1-Speed Automat...        Kia   
35   2024 Ford Mustang Mach-E Premium 300A 99/86 Ci...       Ford   
49   \n      \n        Vehicle pricing includes all...    Hyundai   
..                                                 ...        ...   
884  \n      \n        Heated/Cooled Leather Seats,...        BMW   
893  \n      \n        WHY BUY FROM US?Our dealersh...      Honda   
941  \n      \n        Five Star Hyundai of Macon i...    Hyundai   
944  Yacht Blue 2024 Kia EV6 Wind AWD 1-Speed Autom...        Kia   
978  Aurora Black Pearl 2024 Kia EV6 Wind AWD 1-Spe...        Kia   

              model  year     price  \
14        Blazer EV  2024   51695.0   
28        Blazer EV  2024   52190.0   
33              EV6  2024   49820.0   
35   Mustang Mach-E  2024   47790.0   
49          IONIQ 5  2024   44195.0   
..              ...   ...       ...   
884              i7  2024  195895.0   
893            CR-V  2025   38305.0   
941         IONIQ 5  2024   38201.0   
944             EV6  2024   41528.0   
978             EV6  2024   43439.0   

                                                engine  cylinders      fuel  \
14                                                   c        NaN  Electric   
28                                                   c        NaN  Electric   
33                                                   c        NaN  Electric   
35                                                   c        NaN  Electric   
49                                                   c        NaN  Electric   
..                                                 ...        ...       ...   
884                                                  c        NaN  Electric   
893  d>\n\n    \n    <dt>VIN</dt>\n     7FARS4H71SE...        NaN  Gasoline   
941                                                  c        NaN  Electric   
944                                                  c        NaN  Electric   
978                                                  c        NaN  Electric   

     mileage       transmission               trim   body  doors  \
14       4.0  1-Speed Automatic                2LT    SUV    4.0   
28       6.0  1-Speed Automatic                2LT    SUV    4.0   
33      13.0          Automatic                 GT    SUV    4.0   
35       5.0  1-Speed Automatic            Premium    SUV    4.0   
49      14.0  1-Speed Automatic  SE Standard Range    SUV    4.0   
..       ...                ...                ...    ...    ...   
884      0.0  1-Speed Automatic                M70  Sedan    4.0   
893      0.0      Automatic CVT               EX-L    SUV    NaN   
941     12.0          Automatic                 SE    SUV    4.0   
944     13.0          Automatic               Wind    SUV    4.0   
978     14.0          Automatic               Wind    SUV    4.0   

             exterior_color interior_color        drivetrain  
14   Sterling Gray Metallic          Black   All-wheel Drive  
28              Radiant Red          Black   All-wheel Drive  
33               Yacht Blue          Black   All-wheel Drive  
35               Vapor Blue          Black   All-wheel Drive  
49                    Atlas           Gray  Rear-wheel Drive  
..                      ...     

In [ ]:
df.loc[(df['fuel'] == 'Electric') & (df['cylinders'].isnull()), 'cylinders'] = 0.0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1002 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            1002 non-null   object 
 1   description     1002 non-null   object 
 2   make            1002 non-null   object 
 3   model           1002 non-null   object 
 4   year            1002 non-null   int64  
 5   price           979 non-null    float64
 6   engine          1000 non-null   object 
 7   cylinders       994 non-null    float64
 8   fuel            995 non-null    object 
 9   mileage         968 non-null    float64
 10  transmission    1000 non-null   object 
 11  trim            1001 non-null   object 
 12  body            999 non-null    object 
 13  doors           995 non-null    float64
 14  exterior_color  997 non-null    object 
 15  interior_color  964 non-null    object 
 16  drivetrain      1002 non-null   object 
dtypes: float64(4), int64(1), object(1

In [ ]:
df[df ['cylinders'].isnull()]

name  \
128            2024 Subaru Solterra Premium   
219             2024 Honda Prologue TOURING   
315             2024 Honda Prologue TOURING   
489             2024 Honda Prologue TOURING   
490             2024 Honda Prologue TOURING   
610            2024 Chevrolet Equinox EV RS   
726  2024 Jaguar I-PACE EV400 R-Dynamic HSE   
893                    2025 Honda CR-V EX-L   

                                           description       make       model  \
128  2024 Subaru Solterra Galactic Black Fox Subaru...     Subaru    Solterra   
219  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
315  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
489  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
490  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
610      One of Sonic Automotive's Premier Dealerships  Chevrolet  Equinox EV   
726   Contact our Sales Team at(708)342-0400 for mo...     Jaguar      I-PACE   
893  \n      \n        WHY BUY FROM US?Our dealersh...      Honda        CR-V   

     year    price                                             engine  \
128  2024  39934.0                                                  c   
219  2024  55800.0                                                  c   
315  2024  56550.0                                                  c   
489  2024  55800.0                                                  c   
490  2024  55800.0                                                  c   
610  2024  47495.0               <dt>VIN</dt>\n     3GN7DNRPXRS232327   
726  2024  77053.0  d>\n\n    \n    <dt>VIN</dt>\n     SADHM2S12R1...   
893  2025  38305.0  d>\n\n    \n    <dt>VIN</dt>\n     7FARS4H71SE...   

     cylinders      fuel  mileage       transmission                 trim  \
128        NaN       NaN      5.0  1-Speed Automatic              Premium   
219        NaN       NaN      NaN  1-Speed Automatic              TOURING   
315        NaN       NaN      1.0  1-Speed Automatic              TOURING   
489        NaN       NaN      NaN  1-Speed Automatic              TOURING   
490        NaN       NaN      NaN  1-Speed Automatic              TOURING   
610        NaN       NaN      0.0          Automatic                   RS   
726        NaN       NaN      8.0          Automatic  EV400 R-Dynamic HSE   
893        NaN  Gasoline      0.0      Automatic CVT                 EX-L   

    body  doors          exterior_color interior_color         drivetrain  
128  SUV    4.0                Galactic          Black    All-wheel Drive  
219  SUV    4.0                      BX           Gray    All-wheel Drive  
315  SUV    4.0                    Gray          Black    All-wheel Drive  
489  SUV    4.0                   Pearl          Black    All-wheel Drive  
490  SUV    4.0                    Gray          Black    All-wheel Drive  
610  SUV    NaN  Sterling Gray Metallic          Black  Front-wheel Drive  
726  SUV    5.0         Santorini Black          Ebony    All-wheel Drive  
893  SUV    NaN        Urban Gray Pearl          Black    All-wheel Drive

In [ ]:
df[df['name'] == '2024 Honda Prologue TOURING']

name  \
219  2024 Honda Prologue TOURING   
315  2024 Honda Prologue TOURING   
489  2024 Honda Prologue TOURING   
490  2024 Honda Prologue TOURING   

                                           description   make     model  year  \
219  \n      \n        Shop Your Way Hoffman Honda ...  Honda  Prologue  2024   
315  \n      \n        Shop Your Way Hoffman Honda ...  Honda  Prologue  2024   
489  \n      \n        Shop Your Way Hoffman Honda ...  Honda  Prologue  2024   
490  \n      \n        Shop Your Way Hoffman Honda ...  Honda  Prologue  2024   

       price engine  cylinders fuel  mileage       transmission     trim body  \
219  55800.0      c        NaN  NaN      NaN  1-Speed Automatic  TOURING  SUV   
315  56550.0      c        NaN  NaN      1.0  1-Speed Automatic  TOURING  SUV   
489  55800.0      c        NaN  NaN      NaN  1-Speed Automatic  TOURING  SUV   
490  55800.0      c        NaN  NaN      NaN  1-Speed Automatic  TOURING  SUV   

     doors exterior_color interior_color       drivetrain  
219    4.0             BX           Gray  All-wheel Drive  
315    4.0           Gray          Black  All-wheel Drive  
489    4.0          Pearl          Black  All-wheel Drive  
490    4.0           Gray          Black  All-wheel Drive

In [ ]:
df_nulos = df[df['cylinders'].isnull()]
result = df[df['name'].isin(df_nulos['name'])]

In [ ]:
result

name  \
128            2024 Subaru Solterra Premium   
219             2024 Honda Prologue TOURING   
315             2024 Honda Prologue TOURING   
489             2024 Honda Prologue TOURING   
490             2024 Honda Prologue TOURING   
610            2024 Chevrolet Equinox EV RS   
726  2024 Jaguar I-PACE EV400 R-Dynamic HSE   
893                    2025 Honda CR-V EX-L   

                                           description       make       model  \
128  2024 Subaru Solterra Galactic Black Fox Subaru...     Subaru    Solterra   
219  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
315  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
489  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
490  \n      \n        Shop Your Way Hoffman Honda ...      Honda    Prologue   
610      One of Sonic Automotive's Premier Dealerships  Chevrolet  Equinox EV   
726   Contact our Sales Team at(708)342-0400 for mo...     Jaguar      I-PACE   
893  \n      \n        WHY BUY FROM US?Our dealersh...      Honda        CR-V   

     year    price                                             engine  \
128  2024  39934.0                                                  c   
219  2024  55800.0                                                  c   
315  2024  56550.0                                                  c   
489  2024  55800.0                                                  c   
490  2024  55800.0                                                  c   
610  2024  47495.0               <dt>VIN</dt>\n     3GN7DNRPXRS232327   
726  2024  77053.0  d>\n\n    \n    <dt>VIN</dt>\n     SADHM2S12R1...   
893  2025  38305.0  d>\n\n    \n    <dt>VIN</dt>\n     7FARS4H71SE...   

     cylinders      fuel  mileage       transmission                 trim  \
128        NaN       NaN      5.0  1-Speed Automatic              Premium   
219        NaN       NaN      NaN  1-Speed Automatic              TOURING   
315        NaN       NaN      1.0  1-Speed Automatic              TOURING   
489        NaN       NaN      NaN  1-Speed Automatic              TOURING   
490        NaN       NaN      NaN  1-Speed Automatic              TOURING   
610        NaN       NaN      0.0          Automatic                   RS   
726        NaN       NaN      8.0          Automatic  EV400 R-Dynamic HSE   
893        NaN  Gasoline      0.0      Automatic CVT                 EX-L   

    body  doors          exterior_color interior_color         drivetrain  
128  SUV    4.0                Galactic          Black    All-wheel Drive  
219  SUV    4.0                      BX           Gray    All-wheel Drive  
315  SUV    4.0                    Gray          Black    All-wheel Drive  
489  SUV    4.0                   Pearl          Black    All-wheel Drive  
490  SUV    4.0                    Gray          Black    All-wheel Drive  
610  SUV    NaN  Sterling Gray Metallic          Black  Front-wheel Drive  
726  SUV    5.0         Santorini Black          Ebony    All-wheel Drive  
893  SUV    NaN        Urban Gray Pearl          Black    All-wheel Drive

No entramos datos que nos permitan rellenar los nulos, por lo tanto dropeamos los valores

In [ ]:
df = df.dropna(subset=['cylinders'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 994 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            994 non-null    object 
 1   description     994 non-null    object 
 2   make            994 non-null    object 
 3   model           994 non-null    object 
 4   year            994 non-null    int64  
 5   price           971 non-null    float64
 6   engine          992 non-null    object 
 7   cylinders       994 non-null    float64
 8   fuel            994 non-null    object 
 9   mileage         963 non-null    float64
 10  transmission    992 non-null    object 
 11  trim            993 non-null    object 
 12  body            991 non-null    object 
 13  doors           989 non-null    float64
 14  exterior_color  989 non-null    object 
 15  interior_color  956 non-null    object 
 16  drivetrain      994 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df['interior_color'].value_counts()

interior_color
Black           504
Global Black     84
Gray             76
Jet Black        45
Ebony            42
               ... 
Gray / Black      1
Oyster            1
Tupelo/Black      1
Melange           1
Navy Pier         1
Name: count, Length: 91, dtype: int64

In [ ]:
df[df ['interior_color'].isnull()]

name  \
44                           2024 Acura MDX Advance   
48       2024 Volkswagen Atlas Cross Sport 2.0T SEL   
71                              2024 Acura MDX Base   
77                              2024 Acura MDX Base   
134                   2024 GMC Savana 2500 Work Van   
137                           2024 Acura ZDX A-SPEC   
161                       2024 Acura MDX Technology   
169                   2024 GMC Savana 2500 Work Van   
245                             2024 Acura MDX Base   
286                       2024 Acura MDX Technology   
407                       2024 Acura MDX Technology   
414                       2024 Acura MDX Technology   
416                       2024 Jeep Compass Limited   
443                       2024 Jeep Compass Limited   
446                           2024 GMC Yukon XL SLT   
452                      2024 Kia Sorento Hybrid EX   
454                               2024 Kia Seltos S   
466              2024 Jeep Grand Cherokee L Limited   
478                       2024 Acura MDX Technology   
480                2024 Ford Mustang Mach-E Premium   
582                2024 Ford Mustang Mach-E Premium   
608      2023 Mercedes-Benz Sprinter 2500 High Roof   
612      2024 Mercedes-Benz Sprinter 2500 High Roof   
673                  2024 Honda CR-V Hybrid Sport-L   
675                             2024 Acura MDX Base   
708                               2024 Kia Seltos S   
741    2024 Mercedes-Benz Sprinter 2500 Crew 144 WB   
771                       2024 Ford Maverick Lariat   
790                    2024 Volkswagen Taos 1.5T SE   
812                             2024 Acura MDX Base   
819                        2024 Kia Sportage X-Line   
824                      2024 Jeep Compass Latitude   
844                              2024 Kia Seltos SX   
853                            2024 Dodge Hornet GT   
858          2024 Mercedes-Benz GLA 250 Base 4MATIC   
940  2024 Ford Transit-350 148 WB Medium Roof Cargo   
967                        2024 Kia Sportage X-Line   
971                        2024 Kia Sportage X-Line   

                                           description           make  \
44   Gray 2024 Acura MDX Advance 10-Speed Automatic...          Acura   
48   \n      \n        The installed navigation sys...     Volkswagen   
71   Black 2024 Acura MDX 3.5L 10-Speed Automatic V...          Acura   
77   White 2024 Acura MDX 3.5L 10-Speed Automatic V...          Acura   
134  \n      \n        This 2024 GMC Savana Cargo V...            GMC   
137                                                  -          Acura   
161  White 2024 Acura MDX Technology 10-Speed Autom...          Acura   
169  \n      \n        This 2024 GMC Savana Cargo V...            GMC   
245  Gray 2024 Acura MDX 3.5L 10-Speed Automatic V6...          Acura   
286  Gray 2024 Acura MDX Technology 10-Speed Automa...          Acura   
407  Gray 2024 Acura MDX Technology 10-Speed Automa...          Acura   
414  Gray 2024 Acura MDX Technology 10-Speed Automa...          Acura   
416  \n      \n        Heated Seats, Navigation, Mo...           Jeep   
443  As a premier New Jersey Chrysler, Dodge, FIAT,...           Jeep   
446  \n      \n        Contact Legacy Buick GMC of ...            GMC   
452  \n      \n        2024 Kia Sorento Hybrid EX h...            Kia   
454  \n      \n        MSRP does not include added ...            Kia   
466  \n      \n        Equipment: This unit offers ...           Jeep   
478  Black 2024 Acura MDX Technology 10-Speed Autom...          Acura   
480  2024 Ford Mustang Mach-E Premium Faith's Toyot...           Ford   
582  2024 Ford Mustang Mach-E Premium Faith's Toyot...           Ford   
608  2023 Mercedes-Benz Sprinter 2500 Cargo 170 WB ...  Mercedes-Benz   
612  2024 Mercedes-Benz Sprinter 2500 Cargo 170 WB ...  Mercedes-Benz   
673  \n      \n        This 2024 Honda CR-V Hybrid ...          Honda   
675  Silver 2024 Acura MDX 3.5L 10-Speed Automatic ...          Acura   
708

No tiene sentido desperdiciar datos por no saber el color interior / exterior...

In [ ]:
df['interior_color'] = df['interior_color'].fillna('unknown')

In [ ]:
df['exterior_color'] = df['exterior_color'].fillna('unknown')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 994 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            994 non-null    object 
 1   description     994 non-null    object 
 2   make            994 non-null    object 
 3   model           994 non-null    object 
 4   year            994 non-null    int64  
 5   price           971 non-null    float64
 6   engine          992 non-null    object 
 7   cylinders       994 non-null    float64
 8   fuel            994 non-null    object 
 9   mileage         963 non-null    float64
 10  transmission    992 non-null    object 
 11  trim            993 non-null    object 
 12  body            991 non-null    object 
 13  doors           989 non-null    float64
 14  exterior_color  994 non-null    object 
 15  interior_color  994 non-null    object 
 16  drivetrain      994 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

Los vehículos son nuevos, por lo tanto completamos los valores NaN de mileage con 0.

In [ ]:
df['mileage'] = df['mileage'].fillna(0)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 994 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            994 non-null    object 
 1   description     994 non-null    object 
 2   make            994 non-null    object 
 3   model           994 non-null    object 
 4   year            994 non-null    int64  
 5   price           971 non-null    float64
 6   engine          992 non-null    object 
 7   cylinders       994 non-null    float64
 8   fuel            994 non-null    object 
 9   mileage         994 non-null    float64
 10  transmission    992 non-null    object 
 11  trim            993 non-null    object 
 12  body            991 non-null    object 
 13  doors           989 non-null    float64
 14  exterior_color  994 non-null    object 
 15  interior_color  994 non-null    object 
 16  drivetrain      994 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df_nulos = df[df['price'].isnull()]

In [ ]:
df_nulos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23 entries, 142 to 943
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            23 non-null     object 
 1   description     23 non-null     object 
 2   make            23 non-null     object 
 3   model           23 non-null     object 
 4   year            23 non-null     int64  
 5   price           0 non-null      float64
 6   engine          23 non-null     object 
 7   cylinders       23 non-null     float64
 8   fuel            23 non-null     object 
 9   mileage         23 non-null     float64
 10  transmission    23 non-null     object 
 11  trim            23 non-null     object 
 12  body            23 non-null     object 
 13  doors           23 non-null     float64
 14  exterior_color  23 non-null     object 
 15  interior_color  23 non-null     object 
 16  drivetrain      23 non-null     object 
dtypes: float64(4), int64(1), object(12)
mem

Solo 23 registros son los que no tienen un precio. Para esto tenemos 2 soluciones: podemos rellenar los valores con la media de los demas precios, o eliminar estos registros. Sabiendo que existen esos vehiculos en otros registros, completamos por la media agrupando segun vehículo con mismo nombre, y luego dropeamos los registros restantes.

In [ ]:
precios_promedio = df.groupby('name')['price'].transform('mean')
df['price'] = df['price'].fillna(precios_promedio)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 994 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            994 non-null    object 
 1   description     994 non-null    object 
 2   make            994 non-null    object 
 3   model           994 non-null    object 
 4   year            994 non-null    int64  
 5   price           988 non-null    float64
 6   engine          992 non-null    object 
 7   cylinders       994 non-null    float64
 8   fuel            994 non-null    object 
 9   mileage         994 non-null    float64
 10  transmission    992 non-null    object 
 11  trim            993 non-null    object 
 12  body            991 non-null    object 
 13  doors           989 non-null    float64
 14  exterior_color  994 non-null    object 
 15  interior_color  994 non-null    object 
 16  drivetrain      994 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

De este modo salvamos 17 registros, y dropeamos los restantes.

In [ ]:
df = df.dropna(subset=['price'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 988 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            988 non-null    object 
 1   description     988 non-null    object 
 2   make            988 non-null    object 
 3   model           988 non-null    object 
 4   year            988 non-null    int64  
 5   price           988 non-null    float64
 6   engine          986 non-null    object 
 7   cylinders       988 non-null    float64
 8   fuel            988 non-null    object 
 9   mileage         988 non-null    float64
 10  transmission    986 non-null    object 
 11  trim            987 non-null    object 
 12  body            985 non-null    object 
 13  doors           983 non-null    float64
 14  exterior_color  988 non-null    object 
 15  interior_color  988 non-null    object 
 16  drivetrain      988 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df.trim.value_counts()

trim
Limited                                      69
Base                                         56
Pursuit                                      33
SEL                                          33
Latitude                                     29
                                             ..
quattro                                       1
Premium 55 TFSI e S line quattro S tronic     1
FWD RS                                        1
Longhorn                                      1
SV Intelligent AWD                            1
Name: count, Length: 195, dtype: int64

In [ ]:
df

name  \
0                       2024 Jeep Wagoneer Series II   
1                    2024 Jeep Grand Cherokee Laredo   
2                           2024 GMC Yukon XL Denali   
3                         2023 Dodge Durango Pursuit   
4                              2024 RAM 3500 Laramie   
...                                              ...   
997   2024 Mercedes-Benz Sprinter 2500 Standard Roof   
998           2024 Dodge Hornet Hornet R/T Plus Eawd   
999                          2024 Jeep Wagoneer Base   
1000           2024 Nissan Murano SV Intelligent AWD   
1001                2024 Chevrolet Silverado 2500 WT   

                                            description           make  \
0     \n      \n        Heated Leather Seats, Nav Sy...           Jeep   
1     Al West is committed to offering every custome...           Jeep   
2                                                     -            GMC   
3     White Knuckle Clearcoat 2023 Dodge Durango Pur...          Dodge   
4     \n      \n        2024 Ram 3500 Laramie Billet...            RAM   
...                                                 ...            ...   
997   2024 Mercedes-Benz Sprinter 2500 Cargo 144 WB ...  Mercedes-Benz   
998   Dealer Comments +++ Price Ends 5/31/2024 +++ A...          Dodge   
999   \n      \n        The ALL New Friendship CDJR ...           Jeep   
1000  \n      \n        CVT with Xtronic, AWD.At Tod...         Nissan   
1001  01u 2024 Chevrolet Silverado 2500HD Work Truck...      Chevrolet   

               model  year    price  \
0           Wagoneer  2024  74600.0   
1     Grand Cherokee  2024  50170.0   
2           Yukon XL  2024  96410.0   
3            Durango  2023  46835.0   
4               3500  2024  81663.0   
...              ...   ...      ...   
997    Sprinter 2500  2024  59037.0   
998           Hornet  2024  49720.0   
999         Wagoneer  2024  69085.0   
1000          Murano  2024  43495.0   
1001  Silverado 2500  2024  48995.0   

                                                 engine  cylinders      fuel  \
0                               24V GDI DOHC Twin Turbo        6.0  Gasoline   
1                                                   OHV        6.0  Gasoline   
2     6.2L V-8 gasoline direct injection, variable v...        8.0  Gasoline   
3                                          16V MPFI OHV        8.0  Gasoline   
4                              24V DDI OHV Turbo Diesel        6.0    Diesel   
...                                                 ...        ...       ...   
997                           16V DDI DOHC Turbo Diesel        4.0    Diesel   
998   4 gasoline direct injection, DOHC, Multiair va...        4.0  Gasoline   
999                             24V GDI DOHC Twin Turbo        6.0  Gasoline   
1000  6 DOHC, variable valve control, regular unlead...        6.0  Gasoline   
1001  8 gasoline direct injection, variable valve co...        8.0  Gasoline   

      mileage                         transmission                  trim  \
0        10.0                    8-Speed Automatic             Series II   
1         1.0                    8-Speed Automatic                Laredo   
2         0.0                            Automatic                Denali   
3        32.0                    8-Speed Automatic               Pursuit   
4        10.0                    6-Speed Automatic               Laramie   
...       ...                                  ...                   ...   
997      10.0                    9-Speed Automatic         Standard Roof   
998       0.0  6-Spd Aisin F21-250 PHEV Auto Trans  Hornet R/T Plus Eawd   
999      20.0                    8-Speed Automatic                  Base   
1000      6.0                            Automatic    SV Intelligent AWD   
1001     31.0                            Automatic                    WT   

              body  doors           exterior_color    interior_color  \
0              SUV    4.0                    White      Global Black   

In [ ]:
df[df['trim'].isnull()]

name  \
805  2023 Ford Transit-350    

                                           description  make        model  \
805  Lynn Layton Ford is North Alabama's Fastest Gr...  Ford  Transit-350   

     year    price                   engine  cylinders      fuel  mileage  \
805  2023  54525.0  24V GDI DOHC Twin Turbo        6.0  Gasoline      0.0   

           transmission trim       body  doors exterior_color  \
805  10-Speed Automatic  NaN  Cargo Van    3.0   Oxford White   

        interior_color        drivetrain  
805  Dark Palazzo Gray  Rear-wheel Drive

Es un solo registro, y no tenemos datos suficientes para determinar trim, por lo tanto lo eliminamos.

In [ ]:
df = df.dropna(subset=['trim'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 987 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            987 non-null    object 
 1   description     987 non-null    object 
 2   make            987 non-null    object 
 3   model           987 non-null    object 
 4   year            987 non-null    int64  
 5   price           987 non-null    float64
 6   engine          985 non-null    object 
 7   cylinders       987 non-null    float64
 8   fuel            987 non-null    object 
 9   mileage         987 non-null    float64
 10  transmission    985 non-null    object 
 11  trim            987 non-null    object 
 12  body            984 non-null    object 
 13  doors           982 non-null    float64
 14  exterior_color  987 non-null    object 
 15  interior_color  987 non-null    object 
 16  drivetrain      987 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df.doors.value_counts()

doors
4.0    936
3.0     36
2.0     10
Name: count, dtype: int64

In [ ]:
df[df.doors.isnull()]

name  \
124     2024 Hyundai Santa Fe Calligraphy   
190             2024 Hyundai Santa Fe XRT   
215          2024 Hyundai Santa Fe SE 2.4   
614  2025 Honda CR-V Hybrid Sport Touring   
857             2025 Chevrolet Trax ACTIV   

                                           description       make  \
124  \n      \n        Rockwood Green 2024 Hyundai ...    Hyundai   
190  \n      \n        Shimmering Silver Pearl 2024...    Hyundai   
215  \n      \n        Serenity White 2024 Hyundai ...    Hyundai   
614  \n      \n        !!! CALL US AT (812) 477-620...      Honda   
857  \n      \n        See what's behind you with t...  Chevrolet   

           model  year    price      engine  cylinders      fuel  mileage  \
124     Santa Fe  2024  49905.0  Turbo DOHC        4.0  Gasoline      1.0   
190     Santa Fe  2024  42455.0  Turbo DOHC        4.0  Gasoline      1.0   
215     Santa Fe  2024  38100.0  Turbo DOHC        4.0  Gasoline      1.0   
614  CR-V Hybrid  2025  42150.0         NaN        4.0  Gasoline      1.0   
857         Trax  2025  26190.0         der        6.0  Gasoline      1.0   

                   transmission           trim body  doors     exterior_color  \
124                   Automatic    Calligraphy  SUV    NaN              Green   
190                   Automatic            XRT  SUV    NaN  Shimmering Silver   
215                   Automatic         SE 2.4  SUV    NaN              White   
614  1-Speed CVT with Overdrive  Sport Touring  SUV    NaN             Canyon   
857                   Automatic          ACTIV  SUV    NaN       Summit White   

    interior_color         drivetrain  
124          Black    All-wheel Drive  
190          Black    All-wheel Drive  
215           Gray    All-wheel Drive  
614          Black    All-wheel Drive  
857          Black  Front-wheel Drive

Son todos vehículos SUV, por lo tanto tienen 4 puertas por lógica.

In [ ]:
df.doors = df.doors.fillna(4)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 987 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            987 non-null    object 
 1   description     987 non-null    object 
 2   make            987 non-null    object 
 3   model           987 non-null    object 
 4   year            987 non-null    int64  
 5   price           987 non-null    float64
 6   engine          985 non-null    object 
 7   cylinders       987 non-null    float64
 8   fuel            987 non-null    object 
 9   mileage         987 non-null    float64
 10  transmission    985 non-null    object 
 11  trim            987 non-null    object 
 12  body            984 non-null    object 
 13  doors           987 non-null    float64
 14  exterior_color  987 non-null    object 
 15  interior_color  987 non-null    object 
 16  drivetrain      987 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df[ df.body.isnull()]

name  \
164  2024 Dodge Hornet R/T EAWD   
235  2024 Dodge Hornet R/T EAWD   
687    2024 INFINITI QX50 SPORT   

                                           description      make   model  \
164  \n      \n        Fresh Arrival!, CALL 239-424...     Dodge  Hornet   
235  \n      \n        CALL 239-424-7514 to set up ...     Dodge  Hornet   
687  \n      \n        Radiant White 2024 INFINITI ...  INFINITI    QX50   

     year    price                                             engine  \
164  2024  41497.0  4 gasoline direct injection, DOHC, Multiair va...   
235  2024  41036.0  4 gasoline direct injection, DOHC, Multiair va...   
687  2024  49404.0                                                 ER   

     cylinders      fuel  mileage       transmission      trim body  doors  \
164        4.0  Gasoline     11.0  6-Speed Automatic  R/T EAWD  NaN    4.0   
235        4.0  Gasoline      5.0  6-Speed Automatic  R/T EAWD  NaN    4.0   
687        4.0  Gasoline      7.0    (CVT) CONT VAR.     SPORT  NaN    4.0   

    exterior_color interior_color         drivetrain  
164     Hot Tamale          Black    All-wheel Drive  
235         8 Ball          Black    All-wheel Drive  
687  Radiant White     Monaco Red  Front-wheel Drive

Son sólo 3 registros con body null, buscando por internet determinamos que son de tipo SUV, por lo que se imputa este dato.

In [ ]:
df.body = df.body.fillna('SUV')

In [ ]:
df.body.value_counts()

body
SUV              696
Pickup Truck     154
Sedan             77
Cargo Van         35
Passenger Van     13
Hatchback         10
Convertible        1
Minivan            1
Name: count, dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 987 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            987 non-null    object 
 1   description     987 non-null    object 
 2   make            987 non-null    object 
 3   model           987 non-null    object 
 4   year            987 non-null    int64  
 5   price           987 non-null    float64
 6   engine          985 non-null    object 
 7   cylinders       987 non-null    float64
 8   fuel            987 non-null    object 
 9   mileage         987 non-null    float64
 10  transmission    985 non-null    object 
 11  trim            987 non-null    object 
 12  body            987 non-null    object 
 13  doors           987 non-null    float64
 14  exterior_color  987 non-null    object 
 15  interior_color  987 non-null    object 
 16  drivetrain      987 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df[df.engine.isnull()]

name  \
614  2025 Honda CR-V Hybrid Sport Touring   
803          2024 Jeep Wagoneer Series II   

                                           description   make        model  \
614  \n      \n        !!! CALL US AT (812) 477-620...  Honda  CR-V Hybrid   
803  \n      \n        2024 Jeep Wagoneer Series II...   Jeep     Wagoneer   

     year    price engine  cylinders      fuel  mileage  \
614  2025  42150.0    NaN        4.0  Gasoline      1.0   
803  2024  73999.0    NaN        6.0  Gasoline     59.0   

                   transmission           trim body  doors  \
614  1-Speed CVT with Overdrive  Sport Touring  SUV    4.0   
803           8-Speed Automatic      Series II  SUV    4.0   

             exterior_color interior_color        drivetrain  
614                  Canyon          Black   All-wheel Drive  
803  Bright White Clearcoat          Black  Four-wheel Drive

In [ ]:
df.engine.value_counts()

engine
16V GDI DOHC Turbo                                           128
16V GDI DOHC Turbo Hybrid                                     94
c                                                             90
24V MPFI DOHC                                                 53
16V MPFI OHV                                                  47
                                                            ... 
c ZEV                                                          1
12V GDI OHV                                                    1
6 gasoline direct injection, i-VTEC variable valve contro      1
6 DOHC, VVT variable valve control, engine with cylinder       1
8 gasoline direct injection, variable valve control, regu      1
Name: count, Length: 97, dtype: int64

Al ser solo 2 registros, nuevamente decidimos hacer la imputacion manualmente obteniendo la informacion del sitio oficial que distribuye los vehículos

In [ ]:
df.at[614, 'engine'] = '2.0-liter Atkinson-cycle 4-cylinder'

In [ ]:
df.at[803, 'engine'] = 'Gasoline: 3.0 L Hurricane twin-turbo I6 6.4 L Hemi V8 (2022-2023 Grand Wagoneer) Gasoline mild hybrid (MHEV): 5.7 L Hemi V8 (2022-2023 Wagoneer)'

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 987 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            987 non-null    object 
 1   description     987 non-null    object 
 2   make            987 non-null    object 
 3   model           987 non-null    object 
 4   year            987 non-null    int64  
 5   price           987 non-null    float64
 6   engine          987 non-null    object 
 7   cylinders       987 non-null    float64
 8   fuel            987 non-null    object 
 9   mileage         987 non-null    float64
 10  transmission    985 non-null    object 
 11  trim            987 non-null    object 
 12  body            987 non-null    object 
 13  doors           987 non-null    float64
 14  exterior_color  987 non-null    object 
 15  interior_color  987 non-null    object 
 16  drivetrain      987 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

In [ ]:
df.transmission.value_counts()

transmission
8-Speed Automatic                              320
Automatic                                      234
6-Speed Automatic                               98
1-Speed Automatic                               64
9-Speed Automatic                               61
10-Speed Automatic                              57
Automatic CVT                                   44
8-Speed Automatic with Auto-Shift               17
7-Speed DSGA? Automatic w/ 4MO                  11
8-Speed Automatic with Tiptronic                10
CVT                                              9
6-Spd Aisin F21-250 PHEV Auto Trans              7
8-speed automatic                                5
8-Speed A/T                                      4
7-Speed Automatic with Auto-Shift                4
6-Speed Automatic Electronic with Overdrive      4
9-Speed 948TE Automatic                          4
7-Speed Automatic S tronic                       4
7-Speed DSG Automatic with Tiptronic             4
Variable          

In [ ]:
df[df.transmission.isnull()]

name  \
725          2024 Mercedes-Benz EQS 450 Base 4MATIC   
940  2024 Ford Transit-350 148 WB Medium Roof Cargo   

                                           description           make  \
725  \n      \n        NAV, Sunroof, Heated Leather...  Mercedes-Benz   
940  \n      \n        See what's behind you with t...           Ford   

           model  year     price                      engine  cylinders  \
725      EQS 450  2024  111245.0                           c        0.0   
940  Transit-350  2024   52530.0  24V PDI DOHC Flexible Fuel        6.0   

              fuel  mileage transmission                      trim       body  \
725       Electric     10.0          NaN               Base 4MATIC      Sedan   
940  E85 Flex Fuel      1.0          NaN  148 WB Medium Roof Cargo  Cargo Van   

     doors exterior_color interior_color        drivetrain  
725    4.0          White          Black   All-wheel Drive  
940    3.0          White        unknown  Rear-wheel Drive

Nuevamente, son sólo dos registros

In [ ]:
df.loc[df.name == '2024 Mercedes-Benz EQS 450 Base 4MATIC']

name  \
142  2024 Mercedes-Benz EQS 450 Base 4MATIC   
146  2024 Mercedes-Benz EQS 450 Base 4MATIC   
253  2024 Mercedes-Benz EQS 450 Base 4MATIC   
328  2024 Mercedes-Benz EQS 450 Base 4MATIC   
372  2024 Mercedes-Benz EQS 450 Base 4MATIC   
373  2024 Mercedes-Benz EQS 450 Base 4MATIC   
413  2024 Mercedes-Benz EQS 450 Base 4MATIC   
484  2024 Mercedes-Benz EQS 450 Base 4MATIC   
725  2024 Mercedes-Benz EQS 450 Base 4MATIC   

                                           description           make  \
142  \n      \n        Fletcher Jones Motorcars doe...  Mercedes-Benz   
146  \n      \n        Nav System, Moonroof, Heated...  Mercedes-Benz   
253  \n      \n        Panoramic Roof,Leather Seats...  Mercedes-Benz   
328  \n      \n        Fletcher Jones Motorcars doe...  Mercedes-Benz   
372  \n      \n        Fletcher Jones Motorcars doe...  Mercedes-Benz   
373  \n      \n        Nav System, Moonroof, Heated...  Mercedes-Benz   
413  \n      \n        Manufaktur Alpine Grey,3rd R...  Mercedes-Benz   
484  \n      \n        This Mercedes-Benz EQS boast...  Mercedes-Benz   
725  \n      \n        NAV, Sunroof, Heated Leather...  Mercedes-Benz   

       model  year     price engine  cylinders      fuel  mileage  \
142  EQS 450  2024  113030.0      c        0.0  Electric      8.0   
146  EQS 450  2024  109545.0      c        0.0  Electric     10.0   
253  EQS 450  2024  110395.0      c        0.0  Electric      5.0   
328  EQS 450  2024  113030.0      c        0.0  Electric     10.0   
372  EQS 450  2024  113030.0      c        0.0  Electric      4.0   
373  EQS 450  2024  114850.0      c        0.0  Electric      8.0   
413  EQS 450  2024  114160.0      c        0.0  Electric      5.0   
484  EQS 450  2024  117985.0      c        0.0  Electric      1.0   
725  EQS 450  2024  111245.0      c        0.0  Electric     10.0   

          transmission         trim   body  doors           exterior_color  \
142          Automatic  Base 4MATIC  Sedan    4.0   Graphite Gray Metallic   
146  1-Speed Automatic  Base 4MATIC    SUV    4.0                    Black   
253          Automatic  Base 4MATIC  Sedan    4.0  Obsidian Black Metallic   
328          Automatic  Base 4MATIC  Sedan    4.0  Obsidian Black Metallic   
372          Automatic  Base 4MATIC  Sedan    4.0              Polar White   
373  1-Speed Automatic  Base 4MATIC    SUV    4.0                  unknown   
413  1-Speed Automatic  Base 4MATIC    SUV    4.0              Alpine Gray   
484          Automatic  Base 4MATIC  Sedan    4.0  Obsidian Black Metallic   
725                NaN  Base 4MATIC  Sedan    4.0                    White   

      interior_color       drivetrain  
142             Gray  All-wheel Drive  
146            Black  All-wheel Drive  
253            Black  All-wheel Drive  
328            Black  All-wheel Drive  
372  Macchiato Beige  All-wheel Drive  
373            Black  All-wheel Drive  
413            Black  All-wheel Drive  
484            Black  All-wheel Drive  
725            Black  All-wheel Drive

Observamos que para el mismo modelo, y body tipo Sedan, la trasmission es 'Automatic', por lo tanto se imputa con dicho valor

In [ ]:
df.at[725, 'transmission'] = 'Automatic'

Mismo proceso para el otro vehículo

In [ ]:
df.loc[df.name == '2024 Ford Transit-350 148 WB Medium Roof Cargo']

name  \
940  2024 Ford Transit-350 148 WB Medium Roof Cargo   

                                           description  make        model  \
940  \n      \n        See what's behind you with t...  Ford  Transit-350   

     year    price                      engine  cylinders           fuel  \
940  2024  52530.0  24V PDI DOHC Flexible Fuel        6.0  E85 Flex Fuel   

     mileage transmission                      trim       body  doors  \
940      1.0          NaN  148 WB Medium Roof Cargo  Cargo Van    3.0   

    exterior_color interior_color        drivetrain  
940          White        unknown  Rear-wheel Drive

Es el único registro, buscamos la informacion para imputar el tipo de transmission manualmente.

In [ ]:
df.at[940, 'transmission'] = '10-Speed Automatic'

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 987 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            987 non-null    object 
 1   description     987 non-null    object 
 2   make            987 non-null    object 
 3   model           987 non-null    object 
 4   year            987 non-null    int64  
 5   price           987 non-null    float64
 6   engine          987 non-null    object 
 7   cylinders       987 non-null    float64
 8   fuel            987 non-null    object 
 9   mileage         987 non-null    float64
 10  transmission    987 non-null    object 
 11  trim            987 non-null    object 
 12  body            987 non-null    object 
 13  doors           987 non-null    float64
 14  exterior_color  987 non-null    object 
 15  interior_color  987 non-null    object 
 16  drivetrain      987 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

Finalmente el dataset está libre de valores nulos. De los 1002 registros originales perdimos sólo 15.

#Chequeo de valores duplicados

In [ ]:
len(df[df.duplicated()])

24

Hay 24 registros duplicados, por lo tanto los eliminamos

In [ ]:
df.drop_duplicates(inplace = True)

In [ ]:
len(df[df.duplicated()])

0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 963 entries, 0 to 1001
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            963 non-null    object 
 1   description     963 non-null    object 
 2   make            963 non-null    object 
 3   model           963 non-null    object 
 4   year            963 non-null    int64  
 5   price           963 non-null    float64
 6   engine          963 non-null    object 
 7   cylinders       963 non-null    float64
 8   fuel            963 non-null    object 
 9   mileage         963 non-null    float64
 10  transmission    963 non-null    object 
 11  trim            963 non-null    object 
 12  body            963 non-null    object 
 13  doors           963 non-null    float64
 14  exterior_color  963 non-null    object 
 15  interior_color  963 non-null    object 
 16  drivetrain      963 non-null    object 
dtypes: float64(4), int64(1), object(12)
mem

Finalmente, ya no hay valores duplicados.

In [ ]:
df

name  \
0                       2024 Jeep Wagoneer Series II   
1                    2024 Jeep Grand Cherokee Laredo   
2                           2024 GMC Yukon XL Denali   
3                         2023 Dodge Durango Pursuit   
4                              2024 RAM 3500 Laramie   
...                                              ...   
996                          2024 RAM 2500 Tradesman   
997   2024 Mercedes-Benz Sprinter 2500 Standard Roof   
999                          2024 Jeep Wagoneer Base   
1000           2024 Nissan Murano SV Intelligent AWD   
1001                2024 Chevrolet Silverado 2500 WT   

                                            description           make  \
0     \n      \n        Heated Leather Seats, Nav Sy...           Jeep   
1     Al West is committed to offering every custome...           Jeep   
2                                                     -            GMC   
3     White Knuckle Clearcoat 2023 Dodge Durango Pur...          Dodge   
4     \n      \n        2024 Ram 3500 Laramie Billet...            RAM   
...                                                 ...            ...   
996   Since 1947, Shimkat Motor Company has been the...            RAM   
997   2024 Mercedes-Benz Sprinter 2500 Cargo 144 WB ...  Mercedes-Benz   
999   \n      \n        The ALL New Friendship CDJR ...           Jeep   
1000  \n      \n        CVT with Xtronic, AWD.At Tod...         Nissan   
1001  01u 2024 Chevrolet Silverado 2500HD Work Truck...      Chevrolet   

               model  year    price  \
0           Wagoneer  2024  74600.0   
1     Grand Cherokee  2024  50170.0   
2           Yukon XL  2024  96410.0   
3            Durango  2023  46835.0   
4               3500  2024  81663.0   
...              ...   ...      ...   
996             2500  2024  69315.0   
997    Sprinter 2500  2024  59037.0   
999         Wagoneer  2024  69085.0   
1000          Murano  2024  43495.0   
1001  Silverado 2500  2024  48995.0   

                                                 engine  cylinders      fuel  \
0                               24V GDI DOHC Twin Turbo        6.0  Gasoline   
1                                                   OHV        6.0  Gasoline   
2     6.2L V-8 gasoline direct injection, variable v...        8.0  Gasoline   
3                                          16V MPFI OHV        8.0  Gasoline   
4                              24V DDI OHV Turbo Diesel        6.0    Diesel   
...                                                 ...        ...       ...   
996                            24V DDI OHV Turbo Diesel        6.0    Diesel   
997                           16V DDI DOHC Turbo Diesel        4.0    Diesel   
999                             24V GDI DOHC Twin Turbo        6.0  Gasoline   
1000  6 DOHC, variable valve control, regular unlead...        6.0  Gasoline   
1001  8 gasoline direct injection, variable valve co...        8.0  Gasoline   

      mileage       transmission                trim          body  doors  \
0        10.0  8-Speed Automatic           Series II           SUV    4.0   
1         1.0  8-Speed Automatic              Laredo           SUV    4.0   
2         0.0          Automatic              Denali           SUV    4.0   
3        32.0  8-Speed Automatic             Pursuit           SUV    4.0   
4        10.0  6-Speed Automatic             Laramie  Pickup Truck    4.0   
...       ...                ...                 ...           ...    ...   
996       0.0          Automatic           Tradesman  Pickup Truck    4.0   
997      10.0  9-Speed Automatic       Standard Roof     Cargo Van    3.0   
999      20.0  8-Speed Automatic                Base           SUV    4.0   
1000      6.0          Automatic  SV Intelligent AWD           SUV    4.0   
1001     31.0          Automatic                  WT  Pickup Truck    4.0   

               exterior_color    interior_color        drivetrain  
0                       White      Global Black  Four-wheel Dri

In [ ]:
df['description'] = df['description'].str.strip()

In [ ]:
df.drivetrain.value_counts()

drivetrain
All-wheel Drive      440
Four-wheel Drive     323
Front-wheel Drive    135
Rear-wheel Drive      65
Name: count, dtype: int64

# Creación de Tablas

## Tabla Makers

In [ ]:
# MAKERS TABLE

makers = df[["make"]].drop_duplicates()
makers = makers.sort_values(by="make").reset_index().drop(columns=["index"])
makers

make
0           Acura
1            Audi
2             BMW
3           Buick
4        Cadillac
5       Chevrolet
6        Chrysler
7           Dodge
8            Ford
9             GMC
10        Genesis
11          Honda
12        Hyundai
13       INFINITI
14           Jeep
15            Kia
16     Land Rover
17          Lexus
18        Lincoln
19          Mazda
20  Mercedes-Benz
21         Nissan
22            RAM
23         Subaru
24         Toyota
25     Volkswagen
26          Volvo

## Tabla Engine_Configuration
Se dropea la columna engine por discrepancias en los datos, más de un valor por colummna.

In [ ]:
# ENGINE CONFIGURATION TABLE

# dropeado el campo engine por discrepancias. TODO: dropear filas
engine_conf = df[["cylinders", "fuel", "transmission", "body", "doors", "drivetrain"]].drop_duplicates()
engine_conf = engine_conf.reset_index().drop(columns=["index"])
engine_conf

cylinders      fuel       transmission          body  doors  \
0          6.0  Gasoline  8-Speed Automatic           SUV    4.0   
1          8.0  Gasoline          Automatic           SUV    4.0   
2          8.0  Gasoline  8-Speed Automatic           SUV    4.0   
3          6.0    Diesel  6-Speed Automatic  Pickup Truck    4.0   
4          6.0  Gasoline      Automatic CVT           SUV    4.0   
..         ...       ...                ...           ...    ...   
153        8.0  Gasoline          Automatic  Pickup Truck    2.0   
154        3.0  Gasoline          Automatic           SUV    4.0   
155        4.0  Gasoline          Automatic     Hatchback    4.0   
156        4.0  Gasoline  9-Speed Automatic         Sedan    4.0   
157        8.0  Gasoline          Automatic  Pickup Truck    4.0   

            drivetrain  
0     Four-wheel Drive  
1     Four-wheel Drive  
2      All-wheel Drive  
3     Four-wheel Drive  
4      All-wheel Drive  
..                 ...  
153   Four-wheel Drive  
154  Front-wheel Drive  
155  Front-wheel Drive  
156    All-wheel Drive  
157   Rear-wheel Drive  

[158 rows x 6 columns]

## Tabla Models

In [ ]:
# MODELS TABLE
models_partial = df.drop(columns=["price", "name", "mileage", "description", "interior_color", "exterior_color", "engine"]).drop_duplicates()

# create engine_id foreign key
def find_engine_conf(fields):
    row = engine_conf.loc[
        (engine_conf.cylinders == fields.cylinders) &
        (engine_conf.fuel == fields.fuel) &
        (engine_conf.transmission == fields.transmission) &
        (engine_conf.body == fields.body) &
        (engine_conf.doors == fields.doors) &
        (engine_conf.drivetrain == fields.drivetrain)
    ]

    return engine_conf.index.get_loc(row.index[0])

models = models_partial.copy()
models["engine_id"] = models.apply(lambda row: find_engine_conf(row), axis=1)
models = models.drop(columns=["cylinders", "fuel", "transmission", "body", "doors", "drivetrain"])
models = models.reset_index().drop(columns=["index"])

# create make foreign key
models["make"] = models.apply(lambda row: makers.loc[makers.make == row.make].index[0], axis=1)


models

make           model  year                trim  engine_id
0      14        Wagoneer  2024           Series II          0
1      14  Grand Cherokee  2024              Laredo          0
2       9        Yukon XL  2024              Denali          1
3       7         Durango  2023             Pursuit          2
4      22            3500  2024             Laramie          3
..    ...             ...   ...                 ...        ...
441    12            Kona  2024                  SE         26
442    15        Sportage  2024         SX-Prestige         16
443    15           Forte  2024                 LXS         64
444    21          Murano  2024  SV Intelligent AWD         48
445     5  Silverado 2500  2024                  WT        157

[446 rows x 5 columns]

## Tabla Classifieds

In [ ]:
# CLASSIFIEDS TABLE
c = df.drop(columns=["name", "engine"])

# create model foreign key
def find_model(fields):
    row = models_partial.loc[
        (models_partial.cylinders == fields.cylinders) &
        (models_partial.cylinders == fields.cylinders) &
        (models_partial.fuel == fields.fuel) &
        (models_partial.transmission == fields.transmission) &
        (models_partial.body == fields.body) &
        (models_partial.doors == fields.doors) &
        (models_partial.drivetrain == fields.drivetrain)
    ]
    return models_partial.index.get_loc(row.index[0])

c["model"] = c.apply(lambda row: find_model(row), axis=1)
c = c[["model", "price", "description", "interior_color", "exterior_color", "mileage"]]
c = c.reset_index().drop(columns=["index"])
c

model    price                                        description  \
0        0  74600.0  Heated Leather Seats, Nav System, Moonroof, Th...   
1        0  50170.0  Al West is committed to offering every custome...   
2        2  96410.0                                                  -   
3        3  46835.0  White Knuckle Clearcoat 2023 Dodge Durango Pur...   
4        4  81663.0  2024 Ram 3500 Laramie Billet Silver Metallic C...   
..     ...      ...                                                ...   
958     11  69315.0  Since 1947, Shimkat Motor Company has been the...   
959     85  59037.0  2024 Mercedes-Benz Sprinter 2500 Cargo 144 WB ...   
960      0  69085.0  The ALL New Friendship CDJR of Johnson City of...   
961     74  43495.0  CVT with Xtronic, AWD.At Todd Nissan Of Torrin...   
962    445  48995.0  01u 2024 Chevrolet Silverado 2500HD Work Truck...   

       interior_color           exterior_color  mileage  
0        Global Black                    White     10.0  
1        Global Black                 Metallic      1.0  
2    Teak/Light Shale             Summit White      0.0  
3               Black  White Knuckle Clearcoat     32.0  
4               Black                   Silver     10.0  
..                ...                      ...      ...  
958             Black   Bright White Clearcoat      0.0  
959             Black             Arctic White     10.0  
960             Black            Diamond Black     20.0  
961          Graphite      Pearl White Tricoat      6.0  
962         Jet Black         Wheatland Yellow     31.0  

[963 rows x 6 columns]

# Exportación de las tablas a un csv

In [ ]:
path = 'drive/MyDrive/Dataset_BDD/'

In [ ]:
makers.to_csv(path + 'makers.csv', index=True)

In [ ]:
engine_conf.to_csv(path + 'engine_conf.csv', index=True)

In [ ]:
models.to_csv(path + 'models.csv', index=True)

In [ ]:
c.to_csv(path + 'classifieds.csv', index=True)